### PR TITLE
Add DaemonSet and ClusterRole resources

### PR DESCRIPTION
**Pull request by chrisleck**
Brings in a couple of resources from https://github.com/terraform-providers/terraform-provider-kubernetes/issues/100

@terraform-providers/ecosystem

### CODE REVIEW

File: kubernetes/resource_kubernetes_cluster_role.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+func resourceKubernetesClusterRoleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*kubernetes.Clientset)
+
+	_, name, err := idParts(d.Id())
+	if err != nil {
+		return err
+	}
+	log.Printf("[INFO] Reading cluster role %s", name)
+	cRole, err := conn.RbacV1().ClusterRoles().Get(name, metav1.GetOptions{})
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (chrisleck) [reply kubernetes/resource_kubernetes_cluster_role.go]: Done

File: kubernetes/resource_kubernetes_cluster_role.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	metadata := expandMetadata(d.Get("metadata").([]interface{}))
+	cRole := api.ClusterRole{
+		ObjectMeta: metadata,
+		Rules:      expandClusterRoleRule(d.Get("rule").([]interface{})),
+	}
+
+	log.Printf("[INFO] Updating cluster role %q: %v", name, cRole)
+	out, err := conn.RbacV1().ClusterRoles().Update(&cRole)
```
[the remark this answers is not in the document]
Comment (chrisleck) [reply kubernetes/resource_kubernetes_cluster_role.go]: Done

File: kubernetes/resource_kubernetes_cluster_role.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	}
+	log.Printf("[INFO] Deleting cluster role: %#v", name)
+	err = conn.RbacV1().ClusterRoles().Delete(name, &metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] cluster role %s deleted", name)
+
+	d.SetId("")
```
[the remark this answers is not in the document]
Comment (chrisleck) [reply kubernetes/resource_kubernetes_cluster_role.go]: Nice - done.

File: kubernetes/resource_kubernetes_cluster_role.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+	_, name, err := idParts(d.Id())
+	if err != nil {
+		return false, err
+	}
+
+	log.Printf("[INFO] Checking cluster role %s", name)
+	_, err = conn.RbacV1().ClusterRoles().Get(name, metav1.GetOptions{})
+	if err != nil {
+		if statusErr, ok := err.(*errors.StatusError); ok && statusErr.ErrStatus.Code == 404 {
```
[the remark this answers is not in the document]
Comment (chrisleck) [reply kubernetes/resource_kubernetes_cluster_role.go]: Done

File: kubernetes/resource_kubernetes_cluster_role_test.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+			TestLabelThree = "three"
+		}
+		name = "%s"
+	}
+	rule {
+		api_groups = [""]
+		resources  = ["pods", "pods/log"]
+		verbs = ["get", "list"]
+	}
+}`, name)
```
[the remark this answers is not in the document]
Comment (chrisleck) [reply kubernetes/resource_kubernetes_cluster_role_test.go]: Done

File: kubernetes/resource_kubernetes_daemonset.go [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+							Optional:    true,
+							Default:     10,
+						},
+						"selector": {
+							Type:        schema.TypeList,
+							Description: "A label query over pods that are managed by the DaemonSet.",
+							Optional:    true,
+							ForceNew:    true,
+							MaxItems:    1,
+							Elem: &schema.Resource{
```
[the remark this answers is not in the document]
Comment (chrisleck) [reply kubernetes/resource_kubernetes_daemonset.go]: Done

File: kubernetes/resource_kubernetes_daemonset.go [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+									},
+								},
+							},
+						},
+						"template": {
+							Type:        schema.TypeList,
+							Description: "An object that describes the pod that will be created. The DaemonSet will create exactly one copy of this pod on every node that matches the template's node selector (or on every node if no node selector is specified). More info: https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller#pod-template",
+							Required:    true,
+							MaxItems:    1,
+							Elem: &schema.Resource{
```
[the remark this answers is not in the document]
Comment (chrisleck) [reply kubernetes/resource_kubernetes_daemonset.go]: Done

**Comment by chrisleck** [comment]
Thanks for the review!  All current comments addressed with a new commit.  PTAL when you can.

**Comment by chrisleck** [comment]
@alexsomesan mind taking another look at this when you have a chance?  I've addressed all of your comments.

**Comment by alexsomesan** [comment]
@chrisleck I'll try to review as soon as I get a chance.

**Comment by chrisleck** [comment]
@alexsomesan - pinging to make sure this is still on your radar to review.  Don't want to be pushy, but I do have a project blocked on this :)

**Comment by RobMaskell** [comment]
Just to add my $0.02, I'm not exactly blocked by this but I think I have a bug to report but as I'm currently provisioning the ClusterRole in an unconventional way I wanted to test with this first.

**Comment by alexsomesan** [comment]
@chrisleck I haven't forgotten you! I'll get to this PR shortly. I'm at KubeCon this week so operating at low capacity here.

**Comment by chrisleck** [comment]
@alexsomesan - thanks for the update and enjoy the conference!

**Comment by bukzor** [comment]
I don't have enough context to know whether this matters for this PR, but I went and used the functionality added here and found two missing bits of API. `volume.host_path.type` is unsupported, as well as `volume_mount.mount_propagation`.

1. https://kubernetes.io/docs/concepts/storage/volumes/#hostpath
1. https://kubernetes.io/docs/concepts/storage/volumes/#mount-propagation

**Comment by alexsomesan** [comment]
@chrisleck, are there any acceptance tests for the DaemonSet resource?

**Comment by alexsomesan** [comment]
@chrisleck I'm running into a few small issues trying to test this.
First, I need to make this change to get it to build:
```diff
diff --git a/kubernetes/structures_daemonset.go b/kubernetes/structures_daemonset.go
index c553bf0..dfcae87 100644
--- a/kubernetes/structures_daemonset.go
+++ b/kubernetes/structures_daemonset.go
@@ -77,7 +77,7 @@ func expandDaemonSetSpec(daemonset []interface{}) (appsv1.DaemonSetSpec, error)
        if err != nil {
                return obj, err
        }
-       obj.Template = template
+       obj.Template = *template

        return obj, nil
 }
```

After that, I'm seeing this test failure:
```
~/workspace/terraform-provider-kubernetes(add-daemonset-and-clusterrole*) » TESTARGS="-run '^TestAccKubernetesClusterRole_basic'" make testacc                                        alex@alexs-macbook
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run '^TestAccKubernetesClusterRole_basic' -timeout 120m
?   	github.com/terraform-providers/terraform-provider-kubernetes	[no test files]
=== RUN   TestAccKubernetesClusterRole_basic
--- FAIL: TestAccKubernetesClusterRole_basic (0.46s)
    testing.go:538: Step 1 error: Error applying: 1 error occurred:
        	* kubernetes_cluster_role.test: 1 error occurred:
        	* kubernetes_cluster_role.test: Failed to update ClusterRole: invalid resource name "/tf-acc-test-alex-dv1z7k82d2": [may not contain '/']

FAIL
FAIL	github.com/terraform-providers/terraform-provider-kubernetes/kubernetes	1.182s
make: *** [testacc] Error 1
------------------------------------------------------------
```

**Comment by chrisleck** [comment]
Hmm - not seeing that.  I'll try re-syncing to head and make sure we're clean.

**Comment by chrisleck** [comment]
Ah - turns out I had pushed an older branch :(

**Comment by chrisleck** [comment]
Updated with a fresh sync to master and validated tests are running.  PTAL :)

File: kubernetes/resource_kubernetes_daemonset.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+						"template": {
+							Type:        schema.TypeList,
+							Description: "An object that describes the pod that will be created. The DaemonSet will create exactly one copy of this pod on every node that matches the template's node selector (or on every node if no node selector is specified). More info: https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller#pod-template",
```
Comment (pdecat) [inline kubernetes/resource_kubernetes_daemonset.go]: This description should probably link to the DaemonSet specific 
https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#pod-template documentation or to the more general pod documentation https://kubernetes.io/docs/concepts/workloads/pods/pod-overview/#pod-templates

Comment (chrisleck) [reply kubernetes/resource_kubernetes_daemonset.go]: Done

**Comment by orkaa** [comment]
Hi! Really looking forward to getting this merged since I'm just setting up a new cluster the terraform way. ❤️ 

How soon can I expect this to be released? I'm deciding if I want to wait or develop a workaround in the meantime.

Thanks!

**Comment by jodersky** [comment]
Awesome that this git merged! Is there already a release planned with these new features?

**Comment by ranimufid** [comment]
@alexsomesan, may we please know roughly when the upcoming release with this merged content is going to happen?

**Comment by alexsomesan** [comment]
I was hoping to get more of the pending changes in before a release.

But I can also do a release on Monday since this is something a lot of people are waiting for.

**Comment by ranimufid** [comment]
>But I can also do a release on Monday since this is something a lot of people are waiting for.

That would be greatly appreciated @alexsomesan!

**Comment by alexsomesan** [comment]
Ok, will do :)

**Comment by Natalique** [comment]
> Ok, will do :)

Thanks, can't wait! @alexsomesan

**Comment by yacut** [comment]
@alexsomesan I've already tested it. It works great. Thanks!

**Pull request merged**